libnetdata/arrayalloc/arrayalloc.c

```diff
@@ -0,0 +1,328 @@
+                  ar->element_size, sizeof(uintptr_t), ARAL_NATURAL_ALIGNMENT, ar->internal.element_size, ar->internal.page_ptr_offset);
+
+        //info("ARRAYALLOC: element size %zu, sizeof(uintptr_t) %zu, natural alignment %zu, final element size %zu, page_ptr_offset %zu",
+        //      ar->element_size, sizeof(uintptr_t), ARAL_NATURAL_ALIGNMENT, ar->internal.element_size, ar->internal.page_ptr_offset);
+
+        if (ar->elements < 1000)
+            ar->elements = 1000;
+
+        ar->internal.mmap = (ar->use_mmap && ar->cache_dir && *ar->cache_dir) ? true : false;
+        ar->internal.max_alloc_size = natural_alignment(ar->internal.mmap ? ARAL_MAX_PAGE_SIZE_MMAP : ARAL_MAX_PAGE_SIZE_MALLOC, ARAL_NATURAL_ALIGNMENT);
+
+        if(ar->internal.max_alloc_size % ar->internal.element_size)
+            ar->internal.max_alloc_size -= ar->internal.max_alloc_size % ar->internal.element_size;
+
+        ar->internal.first_page = NULL;
+        ar->internal.last_page = NULL;
+        ar->internal.allocation_multiplier = 1;
+        ar->internal.file_number = 0;
+
+        if(ar->internal.mmap) {
+            char filename[FILENAME_MAX + 1];
+            snprintfz(filename, FILENAME_MAX, "%s/array_alloc.mmap", *ar->cache_dir);
+            int r = mkdir(filename, 0775);
+            if (r != 0 && errno != EEXIST)
+                fatal("Cannot create directory '%s'", filename);
+        }
+
+        ar->internal.initialized = true;
+    }
+
+    netdata_mutex_unlock(&mutex);
+}
+
+static void arrayalloc_free_checks(ARAL *ar, ARAL_FREE *fr) {
+    if(fr->size < ar->internal.element_size)
+        fatal("ARRAYALLOC: free item of size %zu, less than the expected element size %zu", fr->size, ar->internal.element_size);
+
+    if(fr->size % ar->internal.element_size)
+        fatal("ARRAYALLOC: free item of size %zu is not multiple to element size %zu", fr->size, ar->internal.element_size);
+}
+
+void unlink_page(ARAL *ar, ARAL_PAGE *page) {
+    if(unlikely(!page)) return;
+
+    if(page->next)
+        page->next->prev = page->prev;
+
+    if(page->prev)
+        page->prev->next = page->next;
+
+    if(page == ar->internal.first_page)
+        ar->internal.first_page = page->next;
+
+    if(page == ar->internal.last_page)
+        ar->internal.last_page = page->prev;
+}
+
+void link_page_first(ARAL *ar, ARAL_PAGE *page) {
+    page->prev = NULL;
+    page->next = ar->internal.first_page;
+    if(page->next) page->next->prev = page;
+
+    ar->internal.first_page = page;
+
+    if(!ar->internal.last_page)
+        ar->internal.last_page = page;
+}
+
+void link_page_last(ARAL *ar, ARAL_PAGE *page) {
+    page->next = NULL;
+    page->prev = ar->internal.last_page;
+    if(page->prev) page->prev->next = page;
+
+    ar->internal.last_page = page;
+
+    if(!ar->internal.first_page)
+        ar->internal.first_page = page;
+}
+
+ARAL_PAGE *find_page_with_allocation(ARAL *ar, void *ptr) {
```

Comment:
done the first 2.
I don't understand the cast. Can you elaborate?

Reply:
Instead of casting pointers to (size_t) inside the function, you should use the uintptr_t type because you perform comparison and arithmetic operations.